extensions/cache/src/main/java/org/radargun/stages/cache/test/async/Conversations.java

```diff
@@ -0,0 +1,160 @@
+import org.radargun.stages.test.Blackhole;
+import org.radargun.stats.Request;
+import org.radargun.stats.RequestSet;
+import org.radargun.traits.BasicOperations;
+import org.radargun.utils.TimeConverter;
+
+class Conversations {
+   private Conversations() {}
+
+   protected abstract static class RecordingConversationStep implements ConversationStep {
+      public abstract CompletableFuture<?> exec(ConversationContext ctx);
+      public abstract Operation operation();
+
+      @Override
+      public CompletableFuture<?> apply(ConversationContext context) {
+         Request request = context.statistics().startRequest();
+         try {
+            CompletableFuture<?> cf = exec(context);
+            request.requestCompleted();
+            cf.whenComplete((response, ex) -> {
+               if (ex == null) {
+                  request.succeeded(operation());
+                  Blackhole.consume(response);
+               } else {
+                  request.failed(operation());
+               }
+               RequestSet requestSet = context.requestSet();
+               synchronized (requestSet) {
+                  requestSet.add(request);
+               }
+            });
+            return cf;
+         } catch (Exception e) {
+            request.requestFailed();
+            RequestSet requestSet = context.requestSet();
+            synchronized (requestSet) {
+               requestSet.add(request);
+            }
+            CompletableFuture<Object> cf = new CompletableFuture();
+            cf.completeExceptionally(e);
+            return cf;
+         }
+      }
+   }
+
+   @DefinitionElement(name = "get", doc = "cache.get()")
+   public static class Get extends RecordingConversationStep {
+      @Override
+      public CompletableFuture<Object> exec(ConversationContext ctx) {
+         return ctx.cache().get(ctx.provider().getRandomKey(ThreadLocalRandom.current()));
+      }
+
+      @Override
+      public Operation operation() {
+         return BasicOperations.GET;
+      }
+   }
+
+   @DefinitionElement(name = "contains-key", doc = "cache.containsKey()")
+   public static class ContainsKey extends RecordingConversationStep {
+      @Override
+      public CompletableFuture<Boolean> exec(ConversationContext ctx) {
+         return ctx.cache().containsKey(ctx.provider().getRandomKey(ThreadLocalRandom.current()));
+      }
+
+      @Override
+      public Operation operation() {
+         return BasicOperations.CONTAINS_KEY;
+      }
+   }
+
+   @DefinitionElement(name = "put", doc = "cache.put()")
+   public static class Put extends RecordingConversationStep {
+      public CompletableFuture<Void> exec(ConversationContext ctx) {
+         ThreadLocalRandom random = ThreadLocalRandom.current();
+         ContextProvider provider = ctx.provider();
+         Object key = provider.getRandomKey(random);
+         Object value = provider.getRandomValue(random, key);
+         return provider.cache().put(key, value);
```

Comment:
Is there a test/use case where the caches from ContextProvider and ConversationContext can be different? If yes, you should probably get the cache from the ctx in all steps (like in Get step).